server/container/util/src/main/java/org/apache/james/util/StreamUtils.java

```diff
@@ -54,4 +57,12 @@
     public static <T> Stream<T> flatten(Stream<T>... streams) {
         return flatten(Arrays.stream(streams));
     }
+
+    public static <A, B, R> Stream<R> cartesianProduct(Stream<A> firstStream, Stream<B> secondStream, BiFunction<A, B, R> combinator) {
```

Comment:
Seems like you no more uses cartesian product.

If so, we can get a ride of it...